code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/integration/ldap/ad/ADTokenUtils.java

```diff
@@ -14,8 +19,42 @@ public String tokenType() {
 
     @Override
     protected boolean isWhitelisted(List<String> idList) {
-        //TODO Real white listing needed.
-        return true;
+        if (idList == null || idList.isEmpty()) {
+            return false;
+        }
+        List<String> whitelistedValues = fromHashSeparatedString(ADConstants.AD_ALLOWED_IDENTITIES.get());
+
+        for (String id : idList) {
+            for (String whiteId: whitelistedValues){
+                if (StringUtils.equals(id, whiteId)){
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public String toHashSeparatedString(List<Identity> identities) {
+        StringBuilder sb = new StringBuilder();
+        Iterator<Identity> identityIterator = identities.iterator();
+        while (identityIterator.hasNext()){
+            sb.append(identityIterator.next().getId().trim());
+            if (identityIterator.hasNext()) sb.append('#');
+        }
+        return sb.toString();
+    }
+
+    public List<String> fromHashSeparatedString(String string) {
+        if (StringUtils.isEmpty(string)) {
+            return new ArrayList<>();
+        }
+        List<String> strings = new ArrayList<>();
+        String[] splitted = string.split("#");
+        for (String aSplitted : splitted) {
+            String element = aSplitted.trim();
+            strings.add(element);
+        }
+        return strings;
```

Comment:
@alena1108 Added the "#" logic to the Utils class.
Also all "Identity" related code is moved to "ADIdentityProvider"